[Docs] Add Security page to Explore (#8486)
yugabyte/yugabyte-db: docs/content/latest/explore/security/security.md

---
title: Security
headerTitle: Security
linkTitle: Security
description: Overview of security in YugabyteDB.
headcontent: Security in YugabyteDB.
image: /images/section_icons/index/secure.png
aliases:
  - /latest/explore/security/
menu:
  latest:
    identifier: explore-security
    parent: explore
    weight: 930
isTocNested: true
showAsideToc: true
---

Like PostgreSQL, YugabyteDB provides security in multiple ways:

> * **Authentication** - limit access to the database to clients with proper credentials

Reviewer: I'm not sure if it's "client" as opposed to "server", or "client" as a "user"

Author: Reading the Postgresql docs, 'client' appears to refer to whatever is connecting to the database, and a 'user' is defined in the db and given roles etc. A client connects using a database user name. So it is 'client authentication' and 'user authorization'

> * **Authorization** - establish roles and privileges to restrict activities that clients can perform

Reviewer: I'm not sure if it's "client" as opposed to "server", or "client" as a "user"

Author: Fixed

* **Encryption** - encrypt the database, as well as all network communication between servers

* **Auditing** - conduct session- and object-level auditing

> * **Restricting network access** - limit connections to the database using RPC binding

Reviewer 2: If you want to have all your run-in headings parallel...
```suggestion
* **Network access restriction** - limit connections to the database using RPC binding
```

Author: fixed

## Authentication

> Authentication is the process by which the database server establishes the identity of the client, and by extension determines whether the client application (or the user who runs the client application) is permitted to connect with the database user name that was requested. YugabyteDB offers a number of different client authentication methods, including:

Reviewer: Suggestion:
...  including the following:

Author: Fixed

* Password - authenticate using MD5 or SCRAM-SHA-256
* LDAP - use external LDAP services to perform client authentication
* Host-based - authenticate local and remote clients based on IP address and using TLS certificates
* Trust - authorize specific local connections

> The method used to authenticate a particular client connection can be selected on the basis of (client) host address, database, and user.

Reviewer 2: ```suggestion
You can select the method used to authenticate a particular client connection on the basis of (client) host address, database, and user credentials.
```

Author: fixed

> The authentication credentials in YugabyteDB are stored internally in the YB-Master system tables. The authentication mechanisms available to clients depends on what is supported and exposed by the YSQL, YCQL, and YEDIS APIs.

Reviewer 2: ```suggestion
YugabyteDB stores authentication credentials internally in the YB-Master system tables. The authentication mechanisms available to clients depend on what is supported and exposed by the YSQL, YCQL, and YEDIS APIs.
```

Author: fixed

Read more about [how to enable authentication in YugabyteDB](../../secure/authentication).

## Authorization

> YugabyteDB provides role-based access control, consisting of a collection of privileges on resources given to roles. 

Reviewer 2: Since people will be searching for "RBAC"...

```suggestion
YugabyteDB provides role-based access control (RBAC), consisting of a collection of privileges on resources given to roles. 
```

Author: fixed

Read more about [authorization in YugabyteDB](../../secure/authorization).

### Roles

> Roles are essential to implementing and administering access control on a YugabyteDB cluster. Roles can represent individual users or a group of users, and encapsulate a set of privileges that can be assigned to other roles (or users). Roles can be modified to grant users or applications only the essential privileges based on the operations they need to perform against the database. Typically, an administrator role is created first. The administrator then creates additional roles for users.  

Reviewer 2: You're going to get tired of me editing your stuff to remove passive and put it in second person. 🤓

```suggestion
Roles are essential to implementing and administering access control on a YugabyteDB cluster. Roles can represent individual users or a group of users, and encapsulate a set of privileges that can be assigned to other roles (or users). You can modify roles to grant users or applications the minimum required privileges based on the operations they need to perform against the database. Typically, you create an administrator role first, and the administrator then creates additional roles for users.  
```

Author: fixed

### Privileges

> Roles are explicitly granted privileges to access objects in the database using the `GRANT` command. You can, for example, assign read access to one role, data modify access to another role, and alter table access to a third.

Reviewer 2: ```suggestion
You grant privileges explicitly to roles to access objects in the database using the `GRANT` command. You can, for example, assign read access to one role, data modify access to another role, and alter table access to a third.
```

Author: fixed

By default, when an object is created, only the owner has privileges, and other roles must be granted privileges.

> ### Row level access

Reviewer: Row-level access

Author: Fixed

> By default, when an object is created, only the owner has privileges, and other roles must be granted privileges.

Reviewer 2: ```suggestion
By default, only the owner has privileges on new objects; you must grant privileges to other roles explicitly.
```

Author: fixed

In addition to database access permissions available through the `ROLE` and `GRANT` privilege system, YugabyteDB provides a more granular level of security where tables can have row security policies that restrict the rows that users can access.

> Row-level Security (RLS) restricts rows that can be returned by normal queries or inserted, updated, or deleted by DML commands. Row-level security policies can be created specific to a `DML` command or with `ALL` commands. They can also be used to create policies on a particular role or multiple roles.

Reviewer 2: Good job adding the search-bait acronym!

By default, tables do not have any row level policies defined, so that if a user has access privileges to a table, all rows within the table are available to query and update.

> ### Column level access

Reviewer: Column-level access

Author: Fixed

> Column level security in YugabyteDB is used to restrict the users to view a particular column or set of columns in a table. The simplest way to achieve column-level security in YugabyteDB is to create a view that includes only the columns that the user needs to have access to using the `CREATE VIEW` command.

Reviewer: Column-level security

Author: Fixed

> Column-level security in YugabyteDB is used to restrict users to viewing only a particular column or set of columns in a table. You do this by creating a view that includes only the columns that the user needs access to using the `CREATE VIEW` command.

Reviewer 2: ```suggestion
You can use column-level security in YugabyteDB to restrict users to viewing only a particular column or set of columns in a table. You do this by creating a view that includes only the columns that the user needs access to using the `CREATE VIEW` command.
```

Author: fixed

## Encryption

> YugabyteDB supports encryption both of clusters, and all network communication between servers. 

Reviewer 2: I think the first one should be a bit clearer. Do you mean encryption at rest of the data in clusters? Probably worth making this two sentences, as well.

Author: fixed

> ### Encryption in transit

Reviewer 2: Make sure these sub-sections are in the same order as your introductory paragraph under the parent heading...

Author: fixed

[TLS encryption](https://en.wikipedia.org/wiki/Transport_Layer_Security) ensures that network communication between servers is secure. You can configure YugabyteDB to use TLS to encrypt intra-cluster and client to server network communication. Servers are secured using TLS certificates, which can be from a public CA or self-signed.

You should enable encryption in transit for YugabyteDB clusters and clients to ensure privacy and integrity of data transferred over the network.

Read more about enabling [Encryption in transit](../../secure/tls-encryption) in YugabyteDB.

### Encryption at rest

> [Encryption at rest](https://en.wikipedia.org/wiki/Data_at_rest#Encryption) ensures that data at rest, i.e., stored on disk, is protected. You can configure YugabyteDB with a user-generated symmetric key to perform cluster-wide encryption.

Reviewer 2: Don't use e.g. and i.e.!

```suggestion
[Encryption at rest](https://en.wikipedia.org/wiki/Data_at_rest#Encryption) ensures that data at rest (that is, stored on disk) is protected. You can configure YugabyteDB with a user-generated symmetric key to perform cluster-wide encryption.
```

Author: fixed

Read more about enabling [Encryption at rest](../../secure/encryption-at-rest) in YugabyteDB.

> ### Column-level encryption

Reviewer 2: You forgot this one in the introductory paragraph.

Author: fixed

YugabyteDB provides column-level encryption to restrict access to sensitive data such as addresses and credit card details. YugabyteDB uses the PostgreSQL `pgcrypto` extension to enable column level encryption. The `PGP_SYM_ENCRYPT` and `PGP_SYM_DECRYPT` functions of `pgcrypto` are used to encrypt and decrypt column data.

Read more about enabling [column-level encryption](../../secure/column-level-encryption) in YugabyteDB.

## Auditing

> The goal of audit logging is to enable you to produce audit logs often required to comply with government, financial, or ISO certifications. YugabyteDB YSQL uses the PostgreSQL Audit Extension (`pgAudit`) to provide detailed session and/or object audit logging via YugabyteDB TServer logging.

Reviewer 2: This is one of Liza's pet peeves (and I don't particularly like it either!)
```suggestion
The goal of audit logging is to enable you to produce audit logs often required to comply with government, financial, or ISO certifications. YugabyteDB YSQL uses the PostgreSQL Audit Extension (`pgAudit`) to provide detailed session and object audit logging via YugabyteDB TServer logging.
```

Author: reworded

Read more about [audit logging](../../secure/audit-logging) in YugabyteDB.

> ### Session

Reviewer 2: ```suggestion
### Session logging
```

Author: fixed

> Session logging is enabled on a per user session basis. You can enable session logging for all DML and DDL statements and log all relations in DML statements.

Reviewer 2: How do I turn it on? Is that covered in the "audit logging" link?

Author: fixed

> ### Object

Reviewer 2: ```suggestion
### Object logging
```

Author: fixed

Object logging logs statements that affect a particular relation, and is intended to be a finer-grained replacement for session-level logging. As such, it may not make sense to use them in conjunction but one possible scenario would be to use session logging to capture each statement and then supplement that with object logging to get more detail about specific relations.

> In YugabyteDB, object-level audit logging is implemented by reusing the PG role system. The `pgaudit.role` setting defines the role that will be used for audit logging. A relation (TABLE, VIEW, etc.) will be audit logged when the audit role has permissions for the command executed or inherits the permissions from another role. This allows you to effectively have multiple audit roles even though there is a single master role in any context.

Reviewer 2: Etc. is bad too :)

```suggestion
YugabyteDB implements object-level audit logging by reusing the PostgreSQL role system. The `pgaudit.role` setting defines the role that will be used for audit logging. A relation (such as TABLE or VIEW) will be audit logged when the audit role has permissions for the command executed or inherits the permissions from another role. This allows you to effectively have multiple audit roles even though there is a single master role in any context.
```

Author: fixed

## Restricting network access

Ensure that YugabyteDB runs in a trusted network environment, such that:

* Servers running YugabyteDB services are directly accessible only by the servers running the application and database administrators.

> * Only servers running applications can connect to YugabyteDB services on the RPC ports. Access to the [YugabyteDB ports](../../deploy/checklist/#default-ports-reference) should be denied to everybody else.

Reviewer 2: ```suggestion
* Only servers running applications can connect to YugabyteDB services on the RPC ports. Access to the [YugabyteDB ports](../../deploy/checklist/#default-ports-reference) should be denied to all others.
```

Author: fixed

> In addition, you can limit the interfaces on which YugabyteDB instances listen for incoming connections. To specify just the required interfaces when starting `yb-master` and `yb-tserver`, use the `--rpc_bind_addresses` option. Do not bind to the loopback address. Read more in the [Admin Reference](../../reference/configuration/yb-tserver/) section on how to use these options when starting `yb-master` and `yb-tserver` services.

Reviewer 2: ```suggestion
In addition, you can limit the interfaces on which YugabyteDB instances listen for incoming connections. To specify just the required interfaces when starting `yb-master` and `yb-tserver`, use the `--rpc_bind_addresses` option. Do not bind to the loopback address. Refer to the [Admin Reference](../../reference/configuration/yb-tserver/) for more information on using these options.
```

Author: fixed
